basestar-storage-sql/src/main/java/io/basestar/storage/sql/dialect/SnowflakeDialect.java

```diff
@@ -326,4 +328,102 @@ private QueryPart visitFor(final For expression, final boolean all) {
             }
         };
     }
+
+    private void merge(final StringBuilder merge, final org.jooq.Table<?> table, final Field<String> idField) {
+
+        merge.append("MERGE INTO ");
+        merge.append(table.getQualifiedName());
+        merge.append(" AS TARGET USING (?) AS SOURCE ON SOURCE.");
+        merge.append(idField.getUnqualifiedName());
+        merge.append(" = TARGET.");
+        merge.append(idField.getUnqualifiedName());
+        merge.append(" ");
+    }
+
+    private void merge(final StringBuilder merge, final org.jooq.Table<?> table, final Field<String> idField, final Field<Long> versionField) {
+
+        merge(merge, table, idField);
+        merge.append("AND SOURCE.");
+        merge.append(versionField.getUnqualifiedName());
+        merge.append(" = TARGET.");
+        merge.append(versionField.getUnqualifiedName());
+        merge.append(" ");
+    }
+
+    private void mergeNotMatchedInsert(final StringBuilder merge, final Map<Field<?>, SelectField<?>> record) {
+
+        merge.append("WHEN NOT MATCHED THEN INSERT (");
+        merge.append(record.keySet().stream().map(f -> f.getUnqualifiedName().toString()).collect(Collectors.joining(",")));
+        merge.append(") VALUES (");
+        merge.append(record.keySet().stream().map(f -> "SOURCE." + f.getUnqualifiedName().toString()).collect(Collectors.joining(",")));
```

Comment:
might be better to extract the record keys into a list to ensure that the column order is the same in `INSERT()` and `VALUES()`.

Reply:
In practise the order is deterministic (although unpredictable) given any instance of a 'standard' map, but I will change it anyway, and maybe order it explicitly, because there can def. be different orders in different JVMs.